### PR TITLE
Compact ID Tiers [BT-163]

### PR DESCRIPTION
**Pull request by mcovarr**

### CODE REVIEW

**Comment by mcovarr** [comment]
jenkins retest

**Comment by mcovarr** [comment]
jenkins retest

**Comment by mcovarr** [comment]
jenkins retest

**Comment by mcovarr** [comment]
jenkins retest

File: config.json.ctmpl [hunk 1]
```diff
@@ -3,7 +3,10 @@
 {{with $dnsDomain := env "DNS_DOMAIN"}}
 
 {
-    "dataObjectResolutionHost": {{if eq $runContext "fiab"}}"wb-mock-drs-dev.storage.googleapis.com"{{else if eq $environment "prod"}}"gen3.biodatacatalyst.nhlbi.nih.gov"{{else}}"staging.gen3.biodatacatalyst.nhlbi.nih.gov"{{end}},
+    "bioDataCatalystHost": {{if eq $runContext "fiab"}}"wb-mock-drs-dev.storage.googleapis.com"{{else if eq $environment "prod"}}"gen3.biodatacatalyst.nhlbi.nih.gov"{{else}}"staging.gen3.biodatacatalyst.nhlbi.nih.gov"{{end}},
+    "theAnvilHost": {{if eq $runContext "fiab"}}"wb-mock-drs-dev.storage.googleapis.com"{{else if eq $environment "prod"}}"gen3.theanvil.io"{{else}}"staging.theanvil.io"{{end}},
+    "crdcHost": {{if eq $runContext "fiab"}}"wb-mock-drs-dev.storage.googleapis.com"{{else if eq $environment "prod"}}"nci-crdc.datacommons.io"{{else}}"nci-crdc-staging.datacommons.io"{{end}},
+    "kidsFirstHost": {{if eq $runContext "fiab"}}"wb-mock-drs-dev.storage.googleapis.com"{{else if eq $environment "prod"}}"data.kidsfirstdrc.org"{{else}}"gen3staging.kidsfirstdrc.org"{{end}},
```
Comment (kshakir) [inline config.json.ctmpl:9]: All ToL: 

One day this file will go away or at least be simpler. It is a pain to render this file before testing locally, and made it so much harder to debug CI failures, leading eventually to `ENV=mock ...` + `config.js`.

Perhaps the environment variables `ENVIRONMENT`, `RUN_CONTEXT`, and `DNS_DOMAIN` could be exposed somehow to the `config.js`.

Maybe that would just be this existing config file read by `config.js` but rendered containing just:

```go
{{with $environment := env "ENVIRONMENT"}}
{{with $runContext := env "RUN_CONTEXT"}}
{{with $dnsDomain := env "DNS_DOMAIN"}}
{
  "environment": "{{$environment}}",
  "runContext": "{{$runContext}}",
  "dnsDomain": "{{$dnsDomain}}"
}
{{end}}{{end}}{{end}}
```

Or maybe in the multi-repo (firecloud-automated-testing, firecloud-develop, etc.) the environment variables just make it into any docker containers or GCF deployments via standard wired environment variables and not in a file.

But those days do not have to be today.

File: common/config.js [hunk 2]
```diff
@@ -33,13 +43,48 @@ const configDefaults = {
                 default: return `https://broad-bond-${dsdeEnv}.appspot.com`;
             }
         })(),
-    dataObjectResolutionHost:
+    bioDataCatalystHost:
+        (() => {
+            switch (marthaEnv) {
+                case ENV_MOCK:
+                    return HOST_MOCK_DRS;
```
Comment (kshakir) [inline common/config.js]: ToL:

"someone" didn't test their code logic in a previous PR. (Git blame is liar, don't trust it.)

Perhaps this entire bit of logic could be better tested if the export were actually a function that takes in an `env: Object` and returns an `Object`, kind of like `supertest`'s [export](https://github.com/visionmedia/supertest/blob/v6.1.3/index.js#L18).

Then all of the various conditions could be verified via unit tests _before_ they reach those various actual `env` situations, especially production.

I think that is a big amount of testing to pile onto this PR for a previous sin, so marking it ToL and perhaps a future ticket.

**Review by kshakir — APPROVED** [review]

File: config.json.ctmpl (hunk 1 above)
Comment (mcovarr) [reply config.json.ctmpl:9]: Yeah I think I'll procrastinate on cleaning up this one at least until we're sure how we'd like it to work. 🙂

**Comment by mcovarr** [comment]
jenkins retest

File: test/martha/martha_v3.test.js [hunk 3]
```diff
@@ -720,28 +724,28 @@ test('determineDrsType should parse drs:// Data Object uri when host includes a
 test('determineDrsType should parse "dos://" Data Object uri with a host and path', (t) => {
     t.is(
         determineDrsTypeTestWrapper('dos://dg.2345/bar'),
-        `https://${config.dataObjectResolutionHost}/ga4gh/dos/v1/dataobjects/dg.2345/bar`
+        `https://${config.bioDataCatalystHost}/ga4gh/dos/v1/dataobjects/dg.2345/bar`
```
Comment (kshakir) [inline test/martha/martha_v3.test.js]:
```suggestion
        `https://${config.HOST_BIODATA_CATALYST_STAGING}/ga4gh/dos/v1/dataobjects/dg.2345/bar`
```

File: test/martha/martha_v3.test.js [hunk 4]
```diff
@@ -720,28 +724,28 @@ test('determineDrsType should parse drs:// Data Object uri when host includes a
...
     );
 });
 
 test('determineDrsType should parse "drs://" Data Object uri with a host and path', (t) => {
     t.is(
         determineDrsTypeTestWrapper('drs://dg.2345/bar'),
-        `https://${config.dataObjectResolutionHost}/ga4gh/dos/v1/dataobjects/dg.2345/bar`
+        `https://${config.bioDataCatalystHost}/ga4gh/dos/v1/dataobjects/dg.2345/bar`
```
Comment (kshakir) [inline test/martha/martha_v3.test.js]:
```suggestion
        `https://${config.HOST_BIODATA_CATALYST_STAGING}/ga4gh/dos/v1/dataobjects/dg.2345/bar`
```

File: test/martha/martha_v3.test.js [hunk 5]
```diff
@@ -720,28 +724,28 @@ test('determineDrsType should parse drs:// Data Object uri when host includes a
...
     );
 });
 
 test('determineDrsType should parse "drs://dg." Data Object uri with query part', (t) => {
     t.is(
         determineDrsTypeTestWrapper('drs://dg.2345/bar?version=1&bananas=yummy'),
-        `https://${config.dataObjectResolutionHost}/ga4gh/dos/v1/dataobjects/dg.2345/bar?version=1&bananas=yummy`
+        `https://${config.bioDataCatalystHost}/ga4gh/dos/v1/dataobjects/dg.2345/bar?version=1&bananas=yummy`
```
Comment (kshakir) [inline test/martha/martha_v3.test.js]:
```suggestion
        `https://${config.HOST_BIODATA_CATALYST_STAGING}/ga4gh/dos/v1/dataobjects/dg.2345/bar?version=1&bananas=yummy`
```

File: test/martha/martha_v3.test.js [hunk 6]
```diff
@@ -720,28 +724,28 @@ test('determineDrsType should parse drs:// Data Object uri when host includes a
...
     );
 });
 
 test('determineDrsType should parse "drs://" Data Object uri with an expanded host and path', (t) => {
     t.is(
-        determineDrsTypeTestWrapper(`dos://${config.dataObjectResolutionHost}/dg.2345/bar`),
-        `https://${config.dataObjectResolutionHost}/ga4gh/drs/v1/objects/dg.2345/bar`
+        determineDrsTypeTestWrapper(`dos://${config.bioDataCatalystHost}/dg.2345/bar`),
+        `https://${config.bioDataCatalystHost}/ga4gh/drs/v1/objects/dg.2345/bar`
```
Comment (kshakir) [inline test/martha/martha_v3.test.js]:
```suggestion
        determineDrsTypeTestWrapper(`dos://${config.HOST_BIODATA_CATALYST_STAGING}/dg.2345/bar`),
        `https://${config.HOST_BIODATA_CATALYST_STAGING}/ga4gh/drs/v1/objects/dg.2345/bar`
```

**Review by kshakir — COMMENTED** [review]
IMHO now that these fantastic new constants are available, **all** files under `/test` should be updated to use the constant `config.HOST_BIODATA_CATALYST_STAGING` instead of the old variable `config.bioDataCatalystHost`.

However, below I've only added GitHub suggestions for the changes already in this PR.

File: test/common/helpers.test.js [hunk 7]
```diff
@@ -53,24 +53,24 @@ test('dataObjectUriToHttps should parse drs:// Data Object uri when host include
 /**
  * Begin Scenario 2: data objects uri with dg host
  */
-test('dataObjectUriToHttps should parse "dos://dg." Data Object uri to use dataObjectResolutionHost', (t) => {
-    t.is(dataObjectUriToHttps('dos://dg.2345/bar'), `https://${config.dataObjectResolutionHost}/ga4gh/dos/v1/dataobjects/dg.2345/bar`);
+test('dataObjectUriToHttps should parse "dos://dg." Data Object uri to use bioDataCatalystHost', (t) => {
+    t.is(dataObjectUriToHttps('dos://dg.2345/bar'), `https://${config.bioDataCatalystHost}/ga4gh/dos/v1/dataobjects/dg.2345/bar`);
```
Comment (kshakir) [inline test/common/helpers.test.js]:
```suggestion
    t.is(dataObjectUriToHttps('dos://dg.2345/bar'), `https://${config.HOST_BIODATA_CATALYST_STAGING}/ga4gh/dos/v1/dataobjects/dg.2345/bar`);
```

File: test/common/helpers.test.js [hunk 8]
```diff
@@ -53,24 +53,24 @@ test('dataObjectUriToHttps should parse drs:// Data Object uri when host include
...
 });
 
-test('dataObjectUriToHttps should parse "dos://dg." Data Object uri to use dataObjectResolutionHost and preserve case', (t) => {
-    t.is(dataObjectUriToHttps('dos://dg.2345AbCdE/bAr'), `https://${config.dataObjectResolutionHost}/ga4gh/dos/v1/dataobjects/dg.2345AbCdE/bAr`);
+test('dataObjectUriToHttps should parse "dos://dg." Data Object uri to use bioDataCatalystHost and preserve case', (t) => {
+    t.is(dataObjectUriToHttps('dos://dg.2345AbCdE/bAr'), `https://${config.bioDataCatalystHost}/ga4gh/dos/v1/dataobjects/dg.2345AbCdE/bAr`);
```
Comment (kshakir) [inline test/common/helpers.test.js]:
```suggestion
    t.is(dataObjectUriToHttps('dos://dg.2345AbCdE/bAr'), `https://${config.HOST_BIODATA_CATALYST_STAGING}/ga4gh/dos/v1/dataobjects/dg.2345AbCdE/bAr`);
```

File: test/common/helpers.test.js [hunk 9]
```diff
@@ -53,24 +53,24 @@ test('dataObjectUriToHttps should parse drs:// Data Object uri when host include
...
 });
 
-test('dataObjectUriToHttps should parse "drs://dg." Data Object uri to use dataObjectResolutionHost', (t) => {
-    t.is(dataObjectUriToHttps('drs://dg.2345/bar'), `https://${config.dataObjectResolutionHost}/ga4gh/dos/v1/dataobjects/dg.2345/bar`);
+test('dataObjectUriToHttps should parse "drs://dg." Data Object uri to use bioDataCatalystHost', (t) => {
+    t.is(dataObjectUriToHttps('drs://dg.2345/bar'), `https://${config.bioDataCatalystHost}/ga4gh/dos/v1/dataobjects/dg.2345/bar`);
```
Comment (kshakir) [inline test/common/helpers.test.js]:
```suggestion
    t.is(dataObjectUriToHttps('drs://dg.2345/bar'), `https://${config.HOST_BIODATA_CATALYST_STAGING}/ga4gh/dos/v1/dataobjects/dg.2345/bar`);
```

File: test/common/helpers.test.js [hunk 10]
```diff
@@ -53,24 +53,24 @@ test('dataObjectUriToHttps should parse drs:// Data Object uri when host include
...
 });
 
-test('dataObjectUriToHttps should parse "drs://dg." Data Object uri with "/" path to use dataObjectResolutionHost', (t) => {
-    t.is(dataObjectUriToHttps('drs://dg.2345/bar/'), `https://${config.dataObjectResolutionHost}/ga4gh/dos/v1/dataobjects/dg.2345/bar`);
+test('dataObjectUriToHttps should parse "drs://dg." Data Object uri with "/" path to use bioDataCatalystHost', (t) => {
+    t.is(dataObjectUriToHttps('drs://dg.2345/bar/'), `https://${config.bioDataCatalystHost}/ga4gh/dos/v1/dataobjects/dg.2345/bar`);
```
Comment (kshakir) [inline test/common/helpers.test.js]:
```suggestion
    t.is(dataObjectUriToHttps('drs://dg.2345/bar/'), `https://${config.HOST_BIODATA_CATALYST_STAGING}/ga4gh/dos/v1/dataobjects/dg.2345/bar`);
```

File: test/common/helpers.test.js [hunk 11]
```diff
@@ -53,24 +53,24 @@ test('dataObjectUriToHttps should parse drs:// Data Object uri when host include
...
 });
 
-test('dataObjectUriToHttps should parse "drs://dg." Data Object uri with query part to use dataObjectResolutionHost', (t) => {
-    t.is(dataObjectUriToHttps('drs://dg.2345/bar?version=1&bananas=yummy'), `https://${config.dataObjectResolutionHost}/ga4gh/dos/v1/dataobjects/dg.2345/bar?version=1&bananas=yummy`);
+test('dataObjectUriToHttps should parse "drs://dg." Data Object uri with query part to use bioDataCatalystHost', (t) => {
+    t.is(dataObjectUriToHttps('drs://dg.2345/bar?version=1&bananas=yummy'), `https://${config.bioDataCatalystHost}/ga4gh/dos/v1/dataobjects/dg.2345/bar?version=1&bananas=yummy`);
```
Comment (kshakir) [inline test/common/helpers.test.js]:
```suggestion
    t.is(dataObjectUriToHttps('drs://dg.2345/bar?version=1&bananas=yummy'), `https://${config.HOST_BIODATA_CATALYST_STAGING}/ga4gh/dos/v1/dataobjects/dg.2345/bar?version=1&bananas=yummy`);
```

File: test/common/helpers.test.js [hunk 12]
```diff
@@ -98,19 +92,19 @@ test('dataObjectUriToHttps should parse "drs://dg.ANV0" Data Object uri', (t) =>
  * Begin Scenario 3: data objects uri with non-dg host and NO path
  */
 test('should parse "dos://dg." Data Object uri with only a host part without a path', (t) => {
-    t.is(dataObjectUriToHttps('dos://foo-bar-baz'), `https://${config.dataObjectResolutionHost}/ga4gh/dos/v1/dataobjects/foo-bar-baz`);
+    t.is(dataObjectUriToHttps('dos://foo-bar-baz'), `https://${config.bioDataCatalystHost}/ga4gh/dos/v1/dataobjects/foo-bar-baz`);
```
Comment (kshakir) [inline test/common/helpers.test.js]:
```suggestion
    t.is(dataObjectUriToHttps('dos://foo-bar-baz'), `https://${config.HOST_BIODATA_CATALYST_STAGING}/ga4gh/dos/v1/dataobjects/foo-bar-baz`);
```

File: test/common/helpers.test.js [hunk 13]
```diff
@@ -98,19 +92,19 @@ test('dataObjectUriToHttps should parse "drs://dg.ANV0" Data Object uri', (t) =>
...
 });
 
 test('should parse "drs://dg." Data Object uri with only a host part without a path', (t) => {
-    t.is(dataObjectUriToHttps('drs://foo-bar-baz'), `https://${config.dataObjectResolutionHost}/ga4gh/dos/v1/dataobjects/foo-bar-baz`);
+    t.is(dataObjectUriToHttps('drs://foo-bar-baz'), `https://${config.bioDataCatalystHost}/ga4gh/dos/v1/dataobjects/foo-bar-baz`);
```
Comment (kshakir) [inline test/common/helpers.test.js]:
```suggestion
    t.is(dataObjectUriToHttps('drs://foo-bar-baz'), `https://${config.HOST_BIODATA_CATALYST_STAGING}/ga4gh/dos/v1/dataobjects/foo-bar-baz`);
```

File: test/common/helpers.test.js [hunk 14]
```diff
@@ -98,19 +92,19 @@ test('dataObjectUriToHttps should parse "drs://dg.ANV0" Data Object uri', (t) =>
...
 });
 
 test('should parse "drs://dg." Data Object uri with only a host part with a "/" path', (t) => {
-    t.is(dataObjectUriToHttps('drs://foo-bar-baz/'), `https://${config.dataObjectResolutionHost}/ga4gh/dos/v1/dataobjects/foo-bar-baz`);
+    t.is(dataObjectUriToHttps('drs://foo-bar-baz/'), `https://${config.bioDataCatalystHost}/ga4gh/dos/v1/dataobjects/foo-bar-baz`);
```
Comment (kshakir) [inline test/common/helpers.test.js]:
```suggestion
    t.is(dataObjectUriToHttps('drs://foo-bar-baz/'), `https://${config.HOST_BIODATA_CATALYST_STAGING}/ga4gh/dos/v1/dataobjects/foo-bar-baz`);
```

File: test/common/helpers.test.js [hunk 15]
```diff
@@ -98,19 +92,19 @@ test('dataObjectUriToHttps should parse "drs://dg.ANV0" Data Object uri', (t) =>
...
 });
 
 test('should parse "drs://dg." Data Object uri with only a host part with a query part', (t) => {
-    t.is(dataObjectUriToHttps('drs://foo-bar-baz?version=1&bananas=yummy'), `https://${config.dataObjectResolutionHost}/ga4gh/dos/v1/dataobjects/foo-bar-baz?version=1&bananas=yummy`);
+    t.is(dataObjectUriToHttps('drs://foo-bar-baz?version=1&bananas=yummy'), `https://${config.bioDataCatalystHost}/ga4gh/dos/v1/dataobjects/foo-bar-baz?version=1&bananas=yummy`);
```
Comment (kshakir) [inline test/common/helpers.test.js]:
```suggestion
    t.is(dataObjectUriToHttps('drs://foo-bar-baz?version=1&bananas=yummy'), `https://${config.HOST_BIODATA_CATALYST_STAGING}/ga4gh/dos/v1/dataobjects/foo-bar-baz?version=1&bananas=yummy`);
```

**Review by kshakir — APPROVED** [review]
I missed a few non-constants in the tests last round, but otherwise 👍

File: test/_marthaLiveEnv.js [hunk 16]
```diff
@@ -25,27 +27,27 @@ const _publicDataObjectUrlsWithGS = [
 ];
 
 const _publicDataObjectUrlsWithoutGS = [
-    'dos://nci-crdc-staging.datacommons.io/206dfaa6-bcf1-4bc9-b2d0-77179f0f48fc',
+    `dos://${config.HOST_CRDC_STAGING}/206dfaa6-bcf1-4bc9-b2d0-77179f0f48fc`,
```
Comment (breilly2) [inline test/_marthaLiveEnv.js:30]: Use camelCase here?

File: test/common/bond_test.js [hunk 17]
```diff
@@ -68,20 +68,38 @@ test('determineBondProvider should return the AnVIL BondProvider if the URL host
...
 
-test('determineBondProvider should return the dcf-fence BondProvider if the URL host is Kids First', (t) => {
+test('determineBondProvider should return the AnVIL BondProvider if the URL host is the AnVIL staging host', (t) => {
+    t.is(determineBondProvider(`drs://${config.HOST_THE_ANVIL_STAGING}/dg.ANV0/00008531-03d7-418c-b3d3-b7b22b5381a0`), BondProviders.ANVIL);
+});
+
+test('determineBondProvider should return the dcf-fence BondProvider if the URL host is Kids First prod', (t) => {
+    t.is(
+        determineBondProvider(`drs://${config.HOST_KIDS_FIRST_PROD}/ed6be7ab-068e-46c8-824a-f39cfbb885cc`),
+        BondProviders.DCF_FENCE,
```
Comment (breilly2) [inline test/common/bond_test.js:84]: dcf-fence is for CRDC. Is it also used for Kids First? Or are we just asserting the current behavior even if it's not confirmed to be correct? If that's the case, we should put a comment on these tests.

File: package.json [hunk 18]
```diff
@@ -31,6 +31,7 @@
     "minimist": "^1.2.5",
     "nyc": "^15.1.0",
     "sinon": "^9.2.1",
-    "supertest": "^6.0.1"
+    "supertest": "^6.0.1",
+    "tmp": "^0.2.1"
```
Comment (breilly2) [inline package.json:35]: Is this being used or was it an accidental add?

**Review by breilly2 — COMMENTED** [review]
I added some single comments. Done reviewing for now.

File: test/common/config_test.js [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+    t.deepEqual(configJson, {});
+});
+
+test('config parseConfigJson should return an empty json for a missing temp file in ENV_DEV', (t) => {
+    const configPathTmp = tmp.fileSync();
+    fs.writeSync(configPathTmp.fd, '{"hello": "world"}');
+    configPathTmp.removeCallback(); // Delete the file before calling parseConfigJson
+    const configJson = config.parseConfigJson({marthaEnv: config.ENV_DEV, configPath: configPathTmp.name});
+    t.deepEqual(configJson, {});
+});
```
Comment (kshakir) [inline test/common/config_test.js:84]: @breilly2 The `tmp` package is being used in these tests for parsing json files.

File: test/common/bond_test.js (hunk 17 above)
Comment (mcovarr) [reply test/common/bond_test.js:84]: [uhh...](https://github.com/broadinstitute/martha/blob/e1039568bf81c2b7d952d9881c2a9731d954a5ab/common/bond.js#L36)

Comment (mcovarr) [reply test/common/bond_test.js:84]: Actually that link I posted above is a red herring as the `config.HOST_KIDS_FIRST_PROD` URL under test is not a dg URL. `config.HOST_KIDS_FIRST_PROD` [falls all the way through](https://github.com/broadinstitute/martha/blob/e1039568bf81c2b7d952d9881c2a9731d954a5ab/common/bond.js#L59) to the bottom and gets the `default` [DCF_FENCE bond provider](https://github.com/broadinstitute/martha/blob/e1039568bf81c2b7d952d9881c2a9731d954a5ab/common/bond.js#L11).

**Review by breilly2 — APPROVED** [review]
👍

**Pull request merged**